Storage/src/SigningHelper.php

```diff
@@ -0,0 +1,578 @@
+    {
+        $keyFilePath = isset($options['keyFilePath'])
+            ? $options['keyFilePath']
+            : null;
+
+        if ($keyFilePath) {
+            if (!file_exists($keyFilePath)) {
+                throw new \InvalidArgumentException(sprintf(
+                    'Keyfile path %s does not exist.',
+                    $keyFilePath
+                ));
+            }
+
+            $options['keyFile'] = json_decode(file_get_contents($keyFilePath), true);
+            if (json_last_error() !== JSON_ERROR_NONE) {
+                throw new \InvalidArgumentException(sprintf(
+                    'Keyfile path %s does not contain valid json.',
+                    $keyFilePath
+                ));
+            }
+        }
+
+        $rw = $connection->requestWrapper();
+
+        $keyFile = isset($options['keyFile'])
+            ? $options['keyFile']
+            : null;
+        if ($keyFile) {
+            $scopes = isset($options['scopes'])
+                ? $options['scopes']
+                : $rw->scopes();
+
+            $credentials = CredentialsLoader::makeCredentials($scopes, $keyFile);
+        } else {
+            $credentials = $rw->getCredentialsFetcher();
+        }
+
+        if (!($credentials instanceof SignBlobInterface)) {
+            throw new \RuntimeException(sprintf(
+                'Credentials object is of type `%s` and is not valid for signing.',
+                get_class($credentials)
+            ));
+        }
+
+        unset(
+            $options['keyFilePath'],
+            $options['keyFile'],
+            $options['scopes']
+        );
+
+        return [$credentials, $options];
+    }
+
+    /**
+     * Determine the method name to be used for signing.
+     *
+     * @param string|null $version The version name.
+     * @return string
+     * @throws \InvalidArgumentException If the version is not in the valid options.
+     */
+    public static function getSigningMethodName($version)
+    {
+        $version = $version ?: self::DEFAULT_URL_SIGNING_VERSION;
+        switch (strtolower($version)) {
+            case 'v2':
+                $method = 'v2Sign';
+                break;
+
+            case 'v4':
+                $method = 'v4Sign';
+                break;
+
+            default:
+                throw new \InvalidArgumentException('Invalid signing version.');
+        }
+
+        return $method;
+    }
+
+    /**
+     * Sign a URL using Google Signed URLs v2.
+     *
+     * This method will be deprecated in the future.
+     *
+     * @param SignBlobInterface $credentials A credentials instance which
+     *        supports signing strings.
+     * @param Timestamp|\DateTimeInterface|int $expires The signed URL
+     *        expiration.
+     * @param string $resource The URI to the storage resource, preceded by a
+     *        leading slash.
+     * @param string|null $generation The resource generation.
```

Comment:
In the tests this uses an integer, should be `int|null` or perhaps `string|int|null`? (same applies to v4 signature)